File: Data Structures & Algorithms/trapping-rain-water/submission-3.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left, right = 0, len(height) - 1
        leftMax, rightMax = 0, 0
        res = 0
        while left < right:
            if leftMax < rightMax:
                left += 1
                leftMax = max(leftMax, height[left + 1])
                res += leftMax - height[left]
            else:
                right -= 1
                rightMax = max(leftMax, height[right + 1])
                res += leftMax - height[right]
        return res

File: Data Structures & Algorithms/jump-game/submission-3.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int left{0};
        int right{};
        int n{static_cast<int>(nums.size())};
        while (right < n - 1 && left <= right) {
            right = max(right, left + nums[left]);
            left++;
        }
        return right >= n - 1;
    }
};
